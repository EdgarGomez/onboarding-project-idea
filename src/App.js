import React,Â { Fragment, useState, useEffect } from 'react';
import { H1Title, H2Title } from './components/StyledComponents';
import Form from './components/Form';
import Appointment from './components/Appointment';

function App() {

  let initialAppointments = JSON.parse(localStorage.getItem('appointments'));
  if(!initialAppointments) {
    initialAppointments = [];
  }

  const [appointments, saveAppointments] = useState(initialAppointments);

  useEffect( () => {
      let initialAppointments = JSON.parse(localStorage.getItem('appointments'));

      if(initialAppointments) {
        localStorage.setItem('appointments', JSON.stringify(appointments))
      } else {
        localStorage.setItem('appointments', JSON.stringify([]));
      }
  }, [appointments] );

  const createAppointment = appointment => {
    saveAppointments([ ...appointments, appointment ]);
  }

  const deleteAppointment = id => {
     const newAppointments = appointments.filter(appointment => appointment.id !== id );
     saveAppointments(newAppointments);
  }

  const title = appointments.length === 0 ? 'There are no appointments yet.' : 'Manage your Appointments';

  return (
    <Fragment>
      <H1Title data-testid="app-name" data-cy="title">Veterinarian appointments</H1Title>

      <div className="container">
        <div className="row">
          <div className="one-half column">
              <Form 
                createAppointment={createAppointment}
              />
          </div>
          <div className="one-half column">
              <H2Title data-testid="dynamic-title">{title}</H2Title>
              {appointments.map(appointment => (
                <Appointment
                  key={appointment.id}
                  appointment={appointment}
                  deleteAppointment={deleteAppointment}
                />
              ))}
          </div>
        </div>
      </div>
    </Fragment>
  );
}

export default App;
