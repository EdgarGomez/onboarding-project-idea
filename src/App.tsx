import React from 'react';
import { H1Title, H2Title } from './components/StyledComponents';
import Form from './components/Form';
import Appointment from './components/Appointment';
import {Appointment as AppointmentInterface} from './interfaces/interfaces';
import { gql, useQuery } from '@apollo/client';

const APPOINTMENTS = gql`
{
  allAppointments {
    id,
    pet,
    owner,
    date,
    time,
    symptom
  }
}
`;

export const App = () => {
  const { data, loading, error } = useQuery(APPOINTMENTS, {
    pollInterval: 500,
  });

  if (loading) return <H1Title>Loading...</H1Title>;
  if (error) return <pre>{error.message}</pre>

  const allAppointments:AppointmentInterface[]  = data.allAppointments;

  const title: string = allAppointments.length === 0 ? 'There are no appointments yet.' : 'Manage your Appointments';

  return (
    <>
      <H1Title data-testid="app-name" data-cy="title">Veterinarian appointments</H1Title>

      <div className="container">
        <div className="row">
          <div className="one-half column">
              <Form 
              />
          </div>
          <div className="one-half column">
              <H2Title data-testid="dynamic-title">{title}</H2Title>
              {allAppointments.map(appointment => (
                <Appointment
                  key={appointment.id}
                  appointment={appointment}
                />
              ))}
          </div>
        </div>
      </div>
    </>
  );
}
