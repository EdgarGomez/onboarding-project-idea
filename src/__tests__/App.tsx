import React from 'react';
import { render, screen, cleanup, fireEvent } from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect';
import userEvent from '@testing-library/user-event';
import App from '../App';

describe('Testing the App', () => {

    test('<App /> The applications works well the first time.', () => {
        render(<App />);
        expect(screen.getByText('Veterinarian appointments')).toBeInTheDocument();
        expect(screen.getByTestId('app-name').textContent).toBe('Veterinarian appointments');
        expect(screen.getByTestId('app-name').tagName).toBe('H1');

        expect(screen.getByText('Create Appointment')).toBeInTheDocument();
        expect(screen.getByText('There are no appointments yet.')).toBeInTheDocument();
    })

    test('<App /> Check dynamic title.', () => {
        render(<App />);
        userEvent.type(screen.getByTestId('pet'), 'Flash');
        userEvent.type(screen.getByTestId('owner'), 'Edgar');
        userEvent.type(screen.getByTestId('date'), '2021-12-10');
        userEvent.type(screen.getByTestId('time'), '10:30');
        userEvent.type(screen.getByTestId('symptom'), 'Barks but is a cat.');

        const btnSubmit = screen.getByTestId('btn-submit');
        //fireEvent.click(btnSubmit);
        userEvent.click(btnSubmit);

        // using query because maybe exists the alert maybe not, with get inside a conditional rendering will pop an error.
        const alert = screen.queryByTestId('alert');
        expect(alert).not.toBeInTheDocument();

        // Check dynamic title
        expect(screen.getByTestId('dynamic-title').textContent).toBe('Manage your Appointments');
        expect(screen.getByTestId('dynamic-title').textContent).not.toBe('There are no appointments yet.');

    })

    test('<App /> Verify appointments in the DOM', async () => {
        render(<App/>);

        // All "find",or most of them, are synchronous
        const appointments = await screen.findAllByTestId('appointment');

        console.log(appointments.toString());
        // Snapshot creates a file so we can verify its content
        expect(appointments).toMatchSnapshot();

        expect(screen.getByTestId('btn-delete').tagName).toBe('BUTTON');
        expect(screen.getByTestId('btn-delete')).toBeInTheDocument();

        // Verify appointment
        expect(screen.getByText('Flash')).toBeInTheDocument();
    })

    test('<App /> Delete appointment', () => {
        render(<App/>);

        const btnDelete = screen.getByTestId('btn-delete');
        expect(btnDelete.tagName).toBe('BUTTON');
        expect(btnDelete).toBeInTheDocument();

        // Trigger click
        userEvent.click(btnDelete);

        // Button must be already gone
        expect(btnDelete).not.toBeInTheDocument();
        // Appointment must be already gone
        expect(screen.queryByText('Flash')).not.toBeInTheDocument();
        expect(screen.queryByTestId('appointment')).not.toBeInTheDocument();
    })

})