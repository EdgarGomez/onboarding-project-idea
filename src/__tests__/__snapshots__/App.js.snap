// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing the App <App /> Verify appointments in the DOM 1`] = `
Array [
  <div
    class="appointment"
    data-testid="appointment"
  >
    <p>
      Pet: 
      <span>
        Flash
      </span>
       
    </p>
    <p>
      Owner: 
      <span>
        Edgar
      </span>
       
    </p>
    <p>
      Date: 
      <span>
        2021-12-10
      </span>
       
    </p>
    <p>
      Time: 
      <span>
        10:30
      </span>
       
    </p>
    <p>
      Symptom: 
      <span>
        Barks but is a cat.
      </span>
       
    </p>
    <button
      class="button delete u-full-width"
      data-testid="btn-delete"
    >
      Delete Ã—
    </button>
  </div>,
]
`;
